engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -0,0 +1,508 @@
+    }
+  }
+
+  /** Equals for Atoms and AtomConstructors */
+
+  @Specialization
+  boolean equalsAtomConstructors(AtomConstructor selfConstructor, AtomConstructor otherConstructor) {
+    return selfConstructor == otherConstructor;
+  }
+
+  /**
+   * How many {@link EqualsNode} should be created for fields in specialization for atoms.
+   */
+  static final int equalsNodeCountForFields = 10;
+
+  static EqualsNode[] createEqualsNodes(int size) {
+    EqualsNode[] nodes = new EqualsNode[size];
+    Arrays.fill(nodes, EqualsNode.build());
+    return nodes;
+  }
+
+  static InvokeEqualsNode[] createInvokeEqualsNodes(int size) {
+    InvokeEqualsNode[] nodes = new InvokeEqualsNode[size];
+    Arrays.fill(nodes, InvokeEqualsNode.build());
+    return nodes;
+  }
+
+  @Specialization
+  boolean equalsAtoms(Atom self, Atom otherAtom,
+      @Cached LoopConditionProfile loopProfile,
+      @Cached(value = "createEqualsNodes(equalsNodeCountForFields)", allowUncached = true) EqualsNode[] fieldEqualsNodes,
+      @Cached(value = "createInvokeEqualsNodes(equalsNodeCountForFields)", allowUncached = true) InvokeEqualsNode[] fieldInvokeEqualsNodes,
+      @Cached ConditionProfile enoughEqualNodesForFieldsProfile,
+      @Cached ConditionProfile fieldsSizeNotEqualProfile,
+      @Cached ConditionProfile constructorsNotEqualProfile) {
+    if (fieldsSizeNotEqualProfile.profile(
+        self.getFields().length != otherAtom.getFields().length
+    )) {
+      return false;
+    }
+    if (constructorsNotEqualProfile.profile(
+        self.getConstructor() != otherAtom.getConstructor()
+    )) {
+      return false;
+    }
+
+    assert self.getFields().length == otherAtom.getFields().length;
+    int fieldsSize = self.getFields().length;
+    if (enoughEqualNodesForFieldsProfile.profile(fieldsSize <= equalsNodeCountForFields)) {
+      loopProfile.profileCounted(fieldsSize);
+      for (int i = 0; loopProfile.inject(i < fieldsSize); i++) {
+        boolean areFieldsSame;
+        // Note that otherFieldAtom does not have to override `==` method, it is sufficient if
+        // selfFieldAtom overrides `==` method.
+        Object selfField = self.getFields()[i];
+        Object otherField = otherAtom.getFields()[i];
+        if (selfField instanceof Atom selfFieldAtom
+            && otherField instanceof Atom otherFieldAtom
+            && atomOverridesEquals(selfFieldAtom)
+        ) {
+          areFieldsSame =
+              fieldInvokeEqualsNodes[i].execute(selfFieldAtom, otherFieldAtom);
+        } else {
+          areFieldsSame = fieldEqualsNodes[i].execute(
+              selfField,
+              otherField
+          );
+        }
+        if (!areFieldsSame) {
+          return false;
+        }
+      }
+    } else {
+      // If there are more fields than equalsNodeCountForFields, just bailout and use
+      // uncached variant of EqualsNode and InvokeEqualsNode
+      CompilerDirectives.transferToInterpreterAndInvalidate();
```

Comment:
You cannot use `transferToInterpreterAndInvalidate` in my opinion. Use `@TruffleBoundary`.